### PR TITLE
Introduced SourceMap for more verbose error reporting

### PR DESCRIPTION
**Pull request by bdemann**

### CODE REVIEW

File: src/compiler/typescript_to_rust/azle_generate/src/generators/cross_canister_call_functions/mod.rs [hunk 1]
```diff
@@ -38,126 +41,128 @@ struct CrossCanisterCallFunctionNames {
...
+    fn generate_cross_canister_call_functions_infos(&self) -> Vec<CrossCanisterCallFunctionsInfo>;
 }
 
-fn generate_cross_canister_call_functions_infos_from_canister_type_alias_decls(
-    canister_type_alias_decls: &Vec<TsTypeAliasDecl>,
-) -> Vec<CrossCanisterCallFunctionsInfo> {
-    canister_type_alias_decls
-        .iter()
-        .fold(vec![], |acc, canister_type_alias_decl| {
+impl GenerateCrossCanisterCallFunctionsInfos for Vec<AzleTypeAlias<'_>> {
```
Comment (bdemann) [inline src/compiler/typescript_to_rust/azle_generate/src/generators/cross_canister_call_functions/mod.rs:122]: Make sure that all of the names of things that used to be TsTypeAliasDecls are updated to reflect becoming AzleTypeAliases. Make sure that TsTypeAliasDecls are all wrapped in AzleTypeAliases by making sure that they are only directly used in the AzleTypeAlias module

File: src/compiler/typescript_to_rust/azle_generate/src/generators/type_aliases.rs [hunk 2]
```diff
@@ -1,26 +1,28 @@
 use crate::{
     cdk_act::{ActDataType, Actable},
-    ts_ast::ts_type_alias_decl::TsTypeAliasListHelperMethods,
+    ts_ast::{type_alias::azle_type_alias::TsTypeAliasListHelperMethods, AzleTypeAlias},
```
Comment (bdemann) [inline src/compiler/typescript_to_rust/azle_generate/src/generators/type_aliases.rs:3]: Rename TsTypeAliasListHelperMethods to be be AzleTypeAliasListHelperMethods

File: src/compiler/typescript_to_rust/azle_generate/src/ts_ast/ast_traits/get_name.rs [hunk 3]
```diff
@@ -0,0 +1,124 @@
...
+            panic!("{}", self.wrong_number_of_params_error())
+        }
+        match self.params.get(0) {
+            Some(param) => param.get_name(),
+            None => panic!("{}", self.wrong_number_of_params_error()),
+        }
+    }
+}
+
+impl GetName for swc_ecma_ast::TsCallSignatureDecl {
```
Comment (bdemann) [inline src/compiler/typescript_to_rust/azle_generate/src/ts_ast/ast_traits/get_name.rs:112]: Clean up these todos and make it more clear what we actually have todo here

File: src/compiler/typescript_to_rust/azle_generate/src/ts_ast/ast_traits/mod.rs [hunk 4]
```diff
@@ -1,42 +1,67 @@
...
     fn get_ts_type(&self) -> TsType;
 }
 
+impl GetTsType for TsPropertySignature {
+    fn get_ts_type(&self) -> TsType {
+        self.type_ann.as_ref().unwrap().get_ts_type()
+    }
+}
+
+impl GetName for TsFnParam {
```
Comment (bdemann) [inline src/compiler/typescript_to_rust/azle_generate/src/ts_ast/ast_traits/mod.rs:31]: Move this to the TsFnParam file

File: src/compiler/typescript_to_rust/azle_generate/src/ts_ast/program/azle_program.rs [hunk 5]
```diff
@@ -0,0 +1,94 @@
...
+
+    fn get_ast_type_alias_decls(&self) -> Vec<AzleTypeAlias> {
+        match &self.program {
+            Program::Module(module) => module.get_type_alias_decls(&self.source_map),
+            Program::Script(_) => vec![],
+        }
+    }
+}
+
+pub trait TsProgramVecHelperMethods {
```
Comment (bdemann) [inline src/compiler/typescript_to_rust/azle_generate/src/ts_ast/program/azle_program.rs:45]: Rename this to be AzleProgram

File: src/compiler/typescript_to_rust/azle_generate/src/ts_ast/module.rs [hunk 6]
```diff
@@ -1,52 +1,68 @@
...
+        let export_decls: Vec<ExportDecl> = self
+            .body
+            .iter()
+            .filter(|module_item| module_item.is_module_decl())
+            .map(|module_item| module_item.as_module_decl().unwrap().clone())
+            .filter(|module_decl| module_decl.is_export_decl())
+            .map(|module_decl| module_decl.as_export_decl().unwrap().clone())
+            .collect();
+
+        let export_type_alias_decls: Vec<AzleTypeAlias> = export_decls
```
Comment (bdemann) [inline src/compiler/typescript_to_rust/azle_generate/src/ts_ast/module.rs:57]: Rename to export_azle_type_alias

File: src/compiler/typescript_to_rust/azle_generate/src/ts_ast/program/system_canister_method_builder/heartbeat.rs [hunk 7]
```diff
@@ -1,13 +1,11 @@
-use swc_ecma_ast::Program;
-
 use crate::{
     cdk_act::{nodes::ActHeartbeatMethod, CanisterMethodType},
     generators::canister_methods::method_body,
-    ts_ast::program::TsProgramVecHelperMethods,
+    ts_ast::program::{azle_program::TsProgramVecHelperMethods, AzleProgram},
```
Comment (bdemann) [inline src/compiler/typescript_to_rust/azle_generate/src/ts_ast/program/system_canister_method_builder/heartbeat.rs:4]: Make these imports better

File: src/compiler/typescript_to_rust/azle_generate/src/ts_ast/program/system_canister_method_builder/pre_upgrade.rs [hunk 8]
```diff
@@ -1,12 +1,13 @@
 use crate::{
     cdk_act::{nodes::ActPreUpgradeMethod, CanisterMethodType},
     generators::canister_methods::method_body,
-    ts_ast::program::TsProgramVecHelperMethods,
+    ts_ast::program::{azle_program::TsProgramVecHelperMethods, AzleProgram},
```
Comment (bdemann) [inline src/compiler/typescript_to_rust/azle_generate/src/ts_ast/program/system_canister_method_builder/pre_upgrade.rs:4]: fix this import

File: src/compiler/typescript_to_rust/azle_generate/src/ts_ast/ts_ast.rs [hunk 9]
```diff
@@ -59,15 +65,34 @@ impl ToAct for TsAst {
         eprintln!("--- Starting AST to ACT Conversion ------------");
         eprintln!("-----------------------------------------------");
         // Collect AST Information
-        let ast_type_alias_decls = &self.programs.get_ast_type_alias_decls();
+        let ast_type_alias_decls = &self.azle_programs.get_ast_type_alias_decls();
         let ast_canister_type_alias_decls = ast_type_alias_decls.get_ast_ts_canister_decls();
 
+        // let source_map_option = self.azle_programs.iter().find(|azle_program| {
```
Comment (bdemann) [inline src/compiler/typescript_to_rust/azle_generate/src/ts_ast/ts_ast.rs:71]: Get rid of this comment

File: src/compiler/typescript_to_rust/azle_generate/src/ts_ast/ts_tuple_type.rs [hunk 10]
```diff
@@ -37,20 +38,26 @@ impl GetDependencies for TsTupleType {
     }
 }
 
+impl GetString for TsTupleType {
```
Comment (bdemann) [inline src/compiler/typescript_to_rust/azle_generate/src/ts_ast/ts_tuple_type.rs:41]: This shouldn't be in here anymore

File: src/compiler/typescript_to_rust/azle_generate/src/ts_ast/ts_type.rs [hunk 11]
```diff
@@ -46,23 +49,54 @@ impl GetDependencies for TsType {
     }
 }
 
-impl GetName for TsType {
-    fn get_name(&self) -> &str {
+// TODO I am wondering if to_string() might make more sense for these? I just am
```
Comment (bdemann) [inline src/compiler/typescript_to_rust/azle_generate/src/ts_ast/ts_type.rs:52]: Resolve this comment

File: src/compiler/typescript_to_rust/azle_generate/src/ts_ast/ts_type_ref.rs [hunk 12]
```diff
@@ -14,25 +14,33 @@ use std::{
...
+
+trait TsTypeRefPrivateMethods {
+    fn get_enclosed_ts_types(&self) -> Vec<TsType>;
 }
 
 impl GetDependencies for TsTypeRef {
     fn get_dependent_types(
         &self,
-        type_alias_lookup: &HashMap<String, TsTypeAliasDecl>,
+        type_alias_lookup: &HashMap<String, AzleTypeAlias>,
```
Comment (bdemann) [inline src/compiler/typescript_to_rust/azle_generate/src/ts_ast/ts_type_ref.rs:43]: rename this

**Review by lastmjs — CHANGES_REQUESTED** [review]

File: src/compiler/typescript_to_rust/azle_generate/src/cdk_act/generators/ic_object/mod.rs [hunk 13]
```diff
@@ -1,238 +1,246 @@
+use crate::generators::cross_canister_call_functions::CrossCanisterHelperMethods;
+use crate::ts_ast::program::AzleProgram;
```
[the remark this answers is not in the document]
Comment (bdemann) [reply src/compiler/typescript_to_rust/azle_generate/src/cdk_act/generators/ic_object/mod.rs:2]: https://github.com/demergent-labs/azle/issues/725

File: src/compiler/typescript_to_rust/azle_generate/src/cdk_act/generators/ic_object/mod.rs [hunk 14]
```diff
@@ -1,238 +1,246 @@
...
-                0
-            )
-            .property(
-                "_azle_stable_storage",
-                _azle_stable_storage,
-                boa_engine::property::Attribute::all()
-            )
-            .build();
-    }
+pub trait IcObjectHelperMethods {
```
[the remark this answers is not in the document]
Comment (bdemann) [reply src/compiler/typescript_to_rust/azle_generate/src/cdk_act/generators/ic_object/mod.rs:8]: https://github.com/demergent-labs/azle/issues/725

File: src/compiler/typescript_to_rust/azle_generate/src/ts_ast/fn_decl/canister_method_builder.rs [hunk 15]
```diff
@@ -11,12 +12,16 @@ use crate::{
 };
 
 impl CanisterMethodBuilder for FnDecl {
```
[the remark this answers is not in the document]
Comment (bdemann) [reply src/compiler/typescript_to_rust/azle_generate/src/ts_ast/fn_decl/canister_method_builder.rs:14]: Yes we will wrap as we go

**Comment by bdemann** [comment]
Closing if favor of this
https://github.com/demergent-labs/azle/pull/726
which has only the addition of SoruceMaps and AzleTypeAliasDecls

**Pull request closed without merging**